### PR TITLE
servicerouter: select http backend by http header X-Marathon-App-Id

### PR DESCRIPTION
**Pull request by felixb**
you can use the following config to specify backends by http header 'X-MARATHON-APP-ID' with the following config:

HAPROXY_HTTP_FRONTEND_ACL:

    acl app_{cleanedUpHostname} hdr(x-marathon-app-id) -i {appId}
    use_backend {backend} if app_{cleanedUpHostname}

### CODE REVIEW

**Comment by cmaloney** [comment]
Any particular reason to not just hard-code this to always be the header X-Marathon-AppId? That way we end up with one standard way of configuring this, I don't see a lot of value in allowing other arbitrary headers to be used here.

It seems like it should be possible to have this not conflict with the virtual hosts stuff, and instead output two frontend definitions (one for if _VHOST is given, and another for the x-marathon-appid header).

Then this basically becomes a flag to "enable appid routing". I get people might not want to always use vhost and appid routing at the same time, but I can see a number of use cases where it would be handy to have.

**Comment by felixb** [comment]
How about this solution?

File: bin/servicerouter.py [hunk 1]
```diff
@@ -547,17 +577,38 @@ def config(apps, groups):
...
+            if http_appid_frontends:
+                http_appid_frontend_acl = templater.haproxy_http_frontend_appid_acl
+                http_appid_frontends += http_appid_frontend_acl.format(
+                    cleanedUpAppId=cleanedUpAppId,
+                    hostname=app.hostname,
+                    appId=app.appId,
+                    backend=backend
+                )
+
+        if app.mode == 'http':
```
Comment (cmaloney) [inline bin/servicerouter.py]: I don't see what the value of this extra if is, couldn't this statement just be part of the previous `app.mode == 'http'`? Possibly reorder the line up a little bit higher so it isn't after the http_appid_frontends bit for readability.

File: bin/servicerouter.py [hunk 2]
```diff
@@ -83,6 +83,10 @@
     An HTTP frontend that binds to port *:80 by default and gathers
     all virtual hosts as defined by the HAPROXY_{n}_VHOST variable.
 
+  HAPROXY_HTTP_FRONTEND_APPID_HEAD
+    An HTTP frontend that binds to port *:81 by default and gathers
+    all apps in http mode.
```
Comment (cmaloney) [inline bin/servicerouter.py]: Could you add documentation on how an app should use this. Something like:

To use this frontend to forward to your app, configure the app with "HAPROXY_0_MODE=http" then you can access it via  a call to the <servicerouter_host>:81 with the header "X-Marathon-App-Id" set to the Marathon AppId. Note multiple ports being exposed by the same marathon app are not supported


What happens when the same App exposes multiple ports as http? Can we explicitly emit a warning when this happens (and ignore the app entirely) or make the forwarding {app-id}/{port-index}?

**Comment by cmaloney** [comment]
This looks good to me. If you squash down and fix the last pep8 error TravisCI found, I will do a final sanity check and merge it.

**Comment by felixb** [comment]
Squashed and documented.

**Pull request merged**